op-batcher/batcher/driver_test.go

```diff
@@ -0,0 +1,252 @@
+package batcher
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-batcher/batcher/mocks"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum-optimism/optimism/op-node/testlog"
+	"github.com/ethereum-optimism/optimism/op-service/txmgr"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func constructDefaultBatchSubmitter(l log.Logger, mockTxMgr ExternalTxManager, l1Client L1ConfigProvider, l2Client L2ConfigProvider, rollupNode RollupNodeConfigProvider) (*BatchSubmitter, error) {
+	resubmissionTimeout, err := time.ParseDuration("30s")
+	if err != nil {
+		return nil, err
+	}
+
+	pollInterval, err := time.ParseDuration("5s")
+	if err != nil {
+		return nil, err
+	}
+
+	batcherConfig := Config{
+		log:          l,
+		L1Client:     l1Client,
+		L2Client:     l2Client,
+		RollupNode:   rollupNode,
+		PollInterval: pollInterval,
+		TxManagerConfig: txmgr.Config{
+			ResubmissionTimeout:       resubmissionTimeout,
+			ReceiptQueryInterval:      time.Second,
+			NumConfirmations:          1,
+			SafeAbortNonceTooLowCount: 3,
+			From:                      common.Address{},
+			Signer: func(ctx context.Context, from common.Address, tx *types.Transaction) (*types.Transaction, error) {
+				return tx, nil
+			},
+		},
+		From:   common.Address{},
+		Rollup: &rollup.Config{},
+		Channel: ChannelConfig{
+			SeqWindowSize:      15,
+			ChannelTimeout:     40,
+			MaxChannelDuration: 1,
+			SubSafetyMargin:    4,
+			// Set the max frame size to 24 so that we can test sending transactions
+			// The fixed overhead for frame size is 23, so we must be larger or else
+			// the uint64 will underflow, causing the frame size to be essentially unbound
+			MaxFrameSize:     24,
+			TargetFrameSize:  1,
+			TargetNumFrames:  1,
+			ApproxComprRatio: 0.4,
+		},
+	}
+
+	// Construct a tx manager with a mock external tx manager
+	txMgr := TransactionManager{
+		batchInboxAddress: batcherConfig.Rollup.BatchInboxAddress,
+		senderAddress:     batcherConfig.From,
+		chainID:           batcherConfig.Rollup.L1ChainID,
+		txMgr:             mockTxMgr,
+		l1Client:          l1Client,
+		signerFn:          batcherConfig.TxManagerConfig.Signer,
+		log:               l,
+	}
+
+	b := BatchSubmitter{
+		Config: batcherConfig,
+		txMgr:  &txMgr,
+		state:  NewChannelManager(l, batcherConfig.Channel),
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return &b, nil
+}
+
+// TestDriverLoadBlocksIntoState ensures that the [BatchSubmitter] can load blocks into the state.
+func TestDriverLoadBlocksIntoState(t *testing.T) {
```

Comment:
hmm, this is quite a long test to try and read through and the majority of it is setting up mocks which makes it pretty hard to follow and quite dependent on the internal implementation and how it interacts with the mocks.

Given that testing the batcher well depends on being able to control the L1 and L2 chains to setup the different scenarios, I wonder if we should take some time to build suitable stubs or some other kind of library code to make it easier for tests to express what those chains should do.

Maybe we could jump on a call early next week and play around with some options together? I'm not so familiar with this bit of the codebase so not exactly sure what the best options would be but it seems like a good opportunity to improve our testing tools to make this kind of test easier in general.

Reply:
Yea well so this was intended to be as minimal as possible right. Can take offline but introducing a library is a lot of work given that we have an ambitious roadmap to execute on. This will probably be hard enough to get merged, I'd rather spend time and cycles on the most important parts of our roadmap than libraries personally.

Reply:
If we had written the codebase to use proper minimal interfaces, this would have been easy to implement from within `op-e2e`, which is actually where I started writing this pr. But a lot of constructors and internals/unexported funcs and fields take concrete structs and struct pointers - which is impossible to mock in a clean way from an external package like `op-e2e`.

Reply:
Yeah, it's definitely hard to test a lot of this stuff, and coming along afterwards to try and add tests is especially hard. The catch though is if we keep saying it's too hard to write good tests and would take too much time to build the testing infrastructure we need or refactor the code to be more testable, the more untestable code we add and the more time we spend chasing our tails fixing regressions that weren't caught by tests and/or the more time we spend trying to maintain brittle, complicated tests.  Testing well and taking time to refactor is actually what makes you able to go faster.

That said, we may only wind up taking a small step to improving things with this particular test and it winds up looking much like this rather than getting something really clean, but its important that each change we make at least makes things a little better.

Reply:
Had a play with this today to get a better feel for what's possible. I think your refactoring in this PR actually make it a lot easier to get the batcher into the e2e framework and that already has the control we need to be able to create L1 and L2 reorgs.

So I came up with https://github.com/ethereum-optimism/optimism/commit/6814bfae89dffdb84d131f7bc10eb801768de0d8 (the `aj/batcher-e2e` branch) which is still fairly messy but I think is a reasonable start.  It's running the batcher entirely async but using the synchronous miner and sequencer so it's a bit of a weird mix of system and actions test at the moment - that's probably ok, just need to move the file to somewhere sensible and make it clear what we're actually trying to test.

I did need to make a couple of small changes to the batcher itself to make things more deterministic. Instead of calling `Start()` and having it run entirely async, I exposed a `Step()` method which is the code that runs on each tick so now the test can essentially just control when those ticks occur.

In terms of asserting the state of the batcher, we may need to add some accessor functions to `BatchSubmitter` which would be reasonable, or we may be able to hook into the batcher metrics. Using metrics like that is always nice because then the metrics are tested and you know you have enough information being exposed to properly reason about the state of the system - the test is basically doing what a person would do when investigating an issue.

Reply:
Nice! Yes that's the goal - make it much easier to test from within a completely external package like op-e2e.